### ISSUE TITLE
The permit matcher isn't compatible with rails-forward_compatible_controller_tests

### ISSUE THREAD

**Issue by BrianHawley** [thoughtbot/shoulda-matchers#1305]
When you are using the rails-forward_compatible_controller_tests gem to convert your tests to Rails 5 syntax, to prepare to upgrade from Rails < 5, Rails 5 syntax is supported and encouraged for controller tests. There is a `Rails::ForwardCompatibleControllerTests.raise_exception` setting to catch uses of old style syntax in your tests. This setting catches the controller action code generated by the `permit` matcher though, because it goes by the ActionPack version.

If `Shoulda::Matchers::RailsShim.make_controller_request` checks whether `Rails::ForwardCompatibleControllerTests` is defined and generates Rails 5 syntax in that case, `permit` wouldn't generate failing code.

For example, this would fix it:
```ruby
module Shoulda
  module Matchers
    # @private
    module RailsShim
      class << self
        # ...
        def make_controller_request(context, verb, action, request_params)
          params =
            if action_pack_gte_5? || defined?(Rails::ForwardCompatibleControllerTests)
              { params: request_params }
            else
              request_params
            end

          context.__send__(verb, action, params)
        end
        # ...
      end
    end
end
```
the `|| defined?(Rails::ForwardCompatibleControllerTests)` being the fix.

**Comment by mcmire** [comment]
Hi @BrianHawley, sorry for the delay in responding to this. I'm wondering if it might be more helpful to expose some kind of configuration variable instead of having the gem check explicitly for a constant. The reason I say this is that in order to test this, we'd have to add `rails-forward_compatible_controller_tests` as a development dependency and bake it into the integration tests, and I'd rather avoid adding more integration tests if we can given that they take quite a long time to run. If, however, we provide an `attr_accessor` in the Configuration class that allows you to specify whether the controller matchers provided in this gem should operate in Rails 4 mode or Rails 5+ mode, then all we'd have to test for is that.

Basically what I'm proposing is a change to configuration such that you'd say something like this:

``` ruby
Shoulda::Matchers.configure do |config|
  config.use_rails_5_syntax_for_controller_matchers = true
end
```

(Option subject to change.)

Then we could check for this option in `make_controller_request` instead of checking whether Rails::ForwardCompatibleControllerTests was defined. What do you think about this?

**Comment by mcmire** [comment]
I'm going to close this to keep the issues list lean but I'm happy to reopen it if this is still useful.
